### PR TITLE
feat(search): enable deep search with path expressions

### PR DESCRIPTION
**Pull request by patricebender**
support search via path expressions:

- deep search via associations
- include calculated elements if explicitly requested

---

with `cds.infer` being a part of all our query objects, each query comes with `target` and `elements` even without the `infer` function of the `db-service`being called. Hence, it was possible to calculate the search term as a very first step of `cqn4sql`. After the `contains(…)` function has been added to the queries `where` or `having` clause the usual join tree algorithm is executed, making it possible to use path expressions within the search term.

---

rework of https://github.com/cap-js/cds-dbs/pull/252

### CODE REVIEW

File: db-service/test/cqn4sql/search.test.js
Comment (patricebender) [inline db-service/test/cqn4sql/search.test.js:361]: this would not have been possible with existing caching logic

File: db-service/lib/infer/index.js
Comment (johannes-vogel) [inline db-service/lib/infer/index.js:384]: Why do we need to infer the search elements? can't this be part of cqn4sql? I naively expected it there.

Comment (johannes-vogel) [reply db-service/lib/infer/index.js:384]: Can't we do it roughly like this:

1. cqn4sql is called
2. as a very first step SELECT.search is rewritten to search function and appended to where clause
3. infer does its magic
4. cqn4sql builds join tree as usual

If we have an abstract function search that does everything search related, I'd like to keep the rest independent of search. hope it's not too weird and somehow understandable what I'd like to express :) if not, we should have a call :)

Comment (patricebender) [reply db-service/lib/infer/index.js:384]: I will try it.

There a couple of obstacles to overcome if we try to build a search term without inferring the query first.

1. there is no query target, can be solved by relying on Daniels infer by setting the prototype to `SELECT()` if a query has not been initialized with the CQL api.
2. columns are not yet resolved to query elements, hence there is no type information available at the column itself -> this becomes relevant for queries with "aggregations"

Comment (patricebender) [reply db-service/lib/infer/index.js]: okay, I made it work with `cds.infer`. However, I found a bug in Daniels implementation. We need `cap/cds/pull/4160` (hopefully with `cds@7.9.0`) before we can merge this.

Comment (patricebender) [reply db-service/lib/infer/index.js]: unfortunately, after upgrading to cds 7.9, [the HANA job is failing.](https://github.com/cap-js/cds-dbs/actions/runs/9063899197/job/24900854487)

@johannes-vogel @BobdenOs Do these errors look familiar to you? If not, I will have a closer look.

File: db-service/lib/cqn4sql.js
Comment (johannes-vogel) [inline db-service/lib/cqn4sql.js:57]: don't you need to append the search terms instead of replacing the existing where/having clauses?

Comment (patricebender) [reply db-service/lib/cqn4sql.js:57]: This is done in line 243ff

**Comment by patricebender** [comment]
blocked by `cap/issues/issues/16172`

**Comment by johannes-vogel** [comment]
> blocked by `cap/issues/issues/16172` → `cds.infer` can't handle table aliases:
> 
> ```js
> > q = SELECT`from ${Books} as Foo { Foo.ID }`
> Query {
>   SELECT: {
>     from: { ref: [ 'my.bookshop.Books' ], as: 'Foo' },
>     columns: [ { ref: [ 'Foo', 'ID' ] } ]
>   }
> }
> > q.elements
> Uncaught:
> Error: Couldn't resolve element "Foo/ID" in entity my.bookshop.Books [ 'ID', 'title', 'stock' ]
>     at /Users/patricebender/SAPDevelop/foo/node_modules/@sap/cds/lib/ql/infer.js:80:22
>     at Array.forEach (<anonymous>)
>     at Function.elements4 (/Users/patricebender/SAPDevelop/foo/node_modules/@sap/cds/lib/ql/infer.js:61:32)
>     at get elements [as elements] (/Users/patricebender/SAPDevelop/foo/node_modules/@sap/cds/lib/ql/SELECT.js:157:53)
> ```

Why would cds.infer have to handle aliases? Where do they stem from?

**Comment by patricebender** [comment]
> Why would cds.infer have to handle aliases?

My sample illustrates the problem, does it not? A queries column must be resolved to an element which is then attached to the query, this must work with and without table aliases in front of 
a given ref.

> Where do they stem from?
Each `cqn4sql` transformation assigns unique table aliases to the given query. All transformed columns will be absolutified → all refs are prepended with the unique table alias.
This also accounts for where exists subqueries, or correlated subqueries stemming from `expand`

### Why now?

If we want to support path expressions in a search annotation, we must attach a `search` function to the query **before it is put into the join tree algorithm** for processing. But for the calculation of the searchable columns, we need to have an _inferred query at hand_. We need the `target` of a query, to find out which elements must be searched. Hence, I rely on `cds.infer` to do the work instead of doing it by myself. If `cqn4sql` sees an `expand` column, it build a subquery, attaches a table alias and transforms the query. Here the bug I just illustrated kicks in, because `cds.infer` can't resolve table aliases whatsoever.

File: db-service/lib/cqn4sql.js
Comment (johannes-vogel) [inline db-service/lib/cqn4sql.js:230]: renaming and comment are obsolete?

**Review by johannes-vogel — APPROVED** [review]

**Pull request merged**
